### ISSUE TITLE
Add learnable FlexibleSigmoid global epistasis function

### ISSUE THREAD

**Issue by jaredgalloway** [matsengrp/multidms#197]
# Add learnable FlexibleSigmoid global epistasis function

## Summary

Add a `FlexibleSigmoid` global epistasis class with a learnable steepness parameter `k`, enabling the model to adapt the nonlinear transformation's shape to the data rather than relying on the fixed standard logistic. This addresses a systematic bias where the current fixed sigmoid forces α and condition-specific shifts to compensate for shape mismatch, reducing replicate reproducibility of inferred shifts.

### What currently exists

The `multidms` global epistasis module (`jaxmodels.py`) provides two options:

- **`Identity`**: `g(x) = x` — no nonlinearity
- **`Sigmoid`**: `g(x) = expit(x) = 1 / (1 + exp(-x))` — fixed standard logistic

Both are stateless (zero learnable parameters) and marked `static=True` on the `Model` equinox module, meaning they are excluded from gradient-based optimization entirely.

The model predicts functional scores as:

```
predicted_func_score = α * (g(φ(X)) - g(φ(x_wt)))
```

where `φ(X) = β0 + X @ β` is the latent phenotype. The only way to adjust the effective nonlinearity is through `α` (per-condition scaling) and `β0` (per-condition intercept, which shifts the operating point on the sigmoid).

### Why this enhancement is needed

In the SARS-CoV-2 spike DMS analysis with three homologs (Delta, BA.1, BA.2), the fixed sigmoid creates a systematic problem:

1. **Operating point sensitivity**: Conditions whose wildtype falls in the sigmoid's saturating tail (e.g., Delta at `g(φ_wt) ≈ 0.85–0.93`) have low local sensitivity `g'(φ_wt)`. Mutations near wildtype produce weak signal, making shift estimation noisy.

2. **Shape compensation via shifts**: When the true biological nonlinearity is steeper or shallower than the standard logistic, the optimizer absorbs the mismatch into condition-specific shifts — creating spurious shifts that are not reproducible across replicates.

3. **Replicate shift correlation gap**: Delta shift R² between independent replicates is 0.28, versus 0.48 for BA.2 (which sits in the sigmoid's linear regime). This asymmetry suggests the fixed sigmoid shape, not biological signal, drives the gap.

A learnable steepness parameter would allow the sigmoid to adapt its shape, potentially reducing the model's reliance on shifts to absorb nonlinearity mismatch.

### What is missing

- No mechanism to learn the global epistasis function's shape
- The `global_epistasis` field is `static=True`, preventing any gradient flow through GE parameters
- No optimization block exists for GE parameters in the block coordinate descent loop

## Proposed Approach

Add a `FlexibleSigmoid` class with a single learnable parameter `k` (steepness), shared across all conditions:

```
g(x) = expit(k * x) = 1 / (1 + exp(-k * x))
```

When `k = 1`, this reduces exactly to the standard `Sigmoid`. Values `k > 1` produce a steeper transition; `k < 1` produce a shallower one.

**Key design decisions:**

- **One parameter (`k`) only, no midpoint `m`**: A learnable midpoint would be redundant with `β0`, which already shifts each condition's operating point on the curve. Adding `m` would create identifiability issues without adding expressiveness.
- **No learnable asymptotes `L`, `U`**: These would be redundant with `α`, which already scales the output range per condition. Fixing the asymptotes at 0 and 1 keeps the parameterization clean.
- **Shared across conditions**: One `k` value for all conditions. Per-condition steepness would partially duplicate α's role and increase overfitting risk.
- **Softplus parameterization**: Store `log_k` internally and apply `softplus` to ensure `k > 0` and provide unconstrained optimization. Initialize `log_k = log(exp(1) - 1) ≈ 0.5413` so that `k = softplus(log_k) = 1.0` exactly, matching the standard sigmoid at initialization.
- **Optimize in calibration block**: `k` is a global shape parameter analogous to `α`, so it belongs in the calibration block alongside `α` and `logθ`.

### Identifiability with α and β

The steepness `k` interacts with other parameters:

- `k` and `α` are **not** redundant: `α` scales the *output* (functional score space), while `k` changes the *shape* of the nonlinearity (how sharply the sigmoid transitions). A steeper sigmoid with smaller α is different from a shallower sigmoid with larger α — they produce different curvature in the GE landscape.
- `k` and `β` have a scaling interaction: `expit(k * (β0 + X @ β)) = expit(β0' + X @ β')` where `β0' = k*β0` and `β' = k*β`. To break this degeneracy, we **regularize `k` toward 1.0** with a small ridge penalty, anchoring the latent scale and letting `β` values remain interpretable.

The regularization takes the form: `ge_ridge * (k - 1)^2`, added to the calibration block objective.

### Programmatic API

```python
import multidms

# Existing usage (unchanged)
model = multidms.Model(data, ge_type="Sigmoid")
model.fit()

# New flexible sigmoid (ge_ridge defaults to 0.01)
model = multidms.Model(data, ge_type="FlexibleSigmoid")
model.fit()  # ge_ridge=0.01 by default; anchors k toward 1.0

# Access learned steepness after fitting
print(model.steepness)  # e.g., 1.34
```

### Configuration (YAML pipeline)

```yaml
spike:
  fitting:
    ge_type: "FlexibleSigmoid"
    ge_ridge: 0.01
```

### 1. FlexibleSigmoid class

**File:** `multidms/jaxmodels.py` (existing)

Add alongside `Identity` and `Sigmoid`:

```python
class FlexibleSigmoid(GlobalEpistasis):
    r"""Sigmoid with learnable steepness.

    Computes :math:`g(x) = \sigma(k \cdot x)` where :math:`k > 0` is a
    learnable steepness parameter. When :math:`k = 1`, this reduces to
    the standard sigmoid.

    Attributes
    ----------
    log_k : Float[Array, ""]
        Log-steepness (unconstrained parameterization). The effective
        steepness is ``k = softplus(log_k)``.
    """

    # softplus_inverse(1.0) = log(exp(1) - 1) ≈ 0.5413
    # so that k = softplus(log_k) = 1.0 exactly at initialization
    log_k: Float[Array, ""] = jnp.float32(jnp.log(jnp.exp(1.0) - 1.0))

    @property
    def k(self) -> Float[Array, ""]:
        """Effective steepness (always positive)."""
        return jax.nn.softplus(self.log_k)

    def __call__(self, x: Float[Array, ""]) -> Float[Array, ""]:
        r"""Return flexible sigmoid of input."""
        return jax.scipy.special.expit(self.k * x)
```

### 2. Remove static constraint for FlexibleSigmoid

**File:** `multidms/jaxmodels.py` (existing)

The `global_epistasis` field on `Model` is currently `eqx.field(static=True)`. This must be changed so that equinox can differentiate through `FlexibleSigmoid.log_k`, while still treating `Identity` and `Sigmoid` as static (they have no parameters for JAX to trace).

Use `eqx.field(static=False)` (the default) and handle the distinction via the filter_spec:

```python
global_epistasis: GlobalEpistasis = eqx.field(default=Identity())
```

For `Identity` and `Sigmoid` (which have no array leaves), equinox's partition will produce empty parameter sets — no behavioral change. For `FlexibleSigmoid`, `log_k` becomes a trainable leaf.

### 3. Filter_spec updates for all optimization blocks

**File:** `multidms/jaxmodels.py` (existing)

Once `static=True` is removed, `global_epistasis` becomes a pytree node. The filter_spec must provide a matching pytree structure — passing a bare `True` would fail because equinox requires leaf-level booleans matching the model's pytree shape.

Use `jax.tree_util.tree_map` to generate the correct filter structure:

```python
# Calibration block: optimize α, logθ, AND global_epistasis parameters
filter_spec_calibration = Model(
    φ=False,
    α=True,
    logθ=True,
    reference_condition=reference_condition,
    global_epistasis=jax.tree_util.tree_map(lambda _: True, global_epistasis),
)

# β0 block: optimize β0 only, freeze GE parameters
filter_spec_β0 = Model(
    φ={d: Latent(β0=True, β=False) for d in data_sets},
    α=False,
    logθ=False,
    reference_condition=reference_condition,
    global_epistasis=jax.tree_util.tree_map(lambda _: False, global_epistasis),
)
```

For `Identity()` and `Sigmoid()` (which have no array leaves), `tree_map` produces an empty pytree — no behavioral change. For `FlexibleSigmoid`, it maps `log_k` to `True` (calibration) or `False` (β0 block), correctly controlling which blocks optimize `k`.

The β non-bundle and β bundle filter_specs must also set `global_epistasis` to all-`False` using the same `tree_map` pattern.

### 4. GE ridge penalty

**File:** `multidms/jaxmodels.py` (existing)

Add a penalty function:

```python
def _ge_ridge_penalty(model: Model, ge_ridge: float = 0.0) -> Float:
    """Ridge penalty on global epistasis parameters toward defaults.

    For FlexibleSigmoid, penalizes (k - 1)^2 to anchor steepness near
    the standard sigmoid and break the k-β scaling degeneracy.
    For Identity and Sigmoid, returns 0.0 (no parameters to penalize).

    Note: isinstance() on equinox modules is resolved at JAX trace time
    (not runtime), so this branch is safe inside @jax.jit. Do not replace
    with jax.lax.cond.
    """
    if isinstance(model.global_epistasis, FlexibleSigmoid):
        return ge_ridge * (model.global_epistasis.k - 1.0) ** 2
    return 0.0
```

Add `ge_ridge` to the `objective_part` signature and the calibration block `.run()` call:

```python
# Updated signature
def objective_part(trainable, rest, data_sets, scale=1.0,
                   beta0_ridge=0.0, ge_ridge=0.0):
    model = eqx.combine(trainable, rest)
    losses = loss_fn(model, data_sets, **loss_kwargs)
    total = sum(losses.values()) / scale
    total += _beta_ridge_penalty(model, beta0_ridge)
    total += _ge_ridge_penalty(model, ge_ridge)
    return total
```

The `ge_ridge` keyword must be passed **only** to the calibration block's optimizer call (analogous to how `beta0_ridge` is passed only to the β0 block):

```python
# Calibration block — pass ge_ridge
model_calibration, state_calibration = opt_calibration.run(
    model_calibration, model_rest, data_sets, scale=scale,
    ge_ridge=ge_ridge,  # NEW: regularize k toward 1.0
)

# β0 block — pass beta0_ridge only (no ge_ridge)
model_β0, state_β0 = opt_β0.run(
    model_β0, model_rest, data_sets, scale=scale,
    beta0_ridge=beta0_ridge,
)

# β blocks — neither ge_ridge nor beta0_ridge
# (unchanged from current code)
```

Also add `_ge_ridge_penalty` to `objective_total` for consistent convergence checking and initial scale computation:

```python
def objective_total(model, data_sets, scale=1.0,
                    beta0_ridge=0.0, ge_ridge=0.0):
    ...
    total += _ge_ridge_penalty(model, ge_ridge)
    return total
```

Update all `objective_total` call sites (initial scale computation, convergence check) to pass `ge_ridge=ge_ridge`.

### 5. fit() function signature

**File:** `multidms/jaxmodels.py` (existing)

Add `ge_ridge` parameter:

```python
def fit(
    data_sets,
    reference_condition,
    ...
    ge_ridge: Float = 1e-2,  # NEW — default nonzero to prevent k-β degeneracy
    global_epistasis: GlobalEpistasis = Identity(),
    ...
):
```

The default `ge_ridge=1e-2` is nonzero because the identifiability analysis shows that `k` and `β` have a scaling degeneracy that must be broken by regularization. Without it, the optimizer can freely trade off between steeper `k` and smaller `β`, producing unstable fits. The value `1e-2` is chosen to be comparable to `l2reg` in the spike analysis (`1e-2` to `1e-4`), providing a gentle anchor toward `k=1.0` without preventing the model from learning a substantially different steepness when the data supports it. For `Identity` and `Sigmoid`, the penalty evaluates to 0.0 regardless of this setting.

### 6. User-facing Model API

**File:** `multidms/model.py` (existing)

Update `ge_type` validation:

```python
if ge_type not in ["Identity", "Sigmoid", "FlexibleSigmoid"]:
    raise ValueError(...)
```

Update the mapping in `fit()`:

```python
if self._ge_type == "Identity":
    global_epistasis = jaxmodels.Identity()
elif self._ge_type == "Sigmoid":
    global_epistasis = jaxmodels.Sigmoid()
elif self._ge_type == "FlexibleSigmoid":
    global_epistasis = jaxmodels.FlexibleSigmoid()
```

Add a `steepness` property:

```python
@property
def steepness(self) -> float | None:
    """Learned steepness parameter (FlexibleSigmoid only)."""
    if self._jax_model is None:
        return None
    ge = self._jax_model.global_epistasis
    if isinstance(ge, jaxmodels.FlexibleSigmoid):
        return float(ge.k)
    return None
```

Add `ge_ridge` to `fit()` kwargs and pass through to `jaxmodels.fit()`.

### JAX tracing considerations

The key change is removing `static=True` from the `global_epistasis` field. This means JAX will now trace through `global_epistasis.__call__` during gradient computation. For `Identity` and `Sigmoid`, this is safe — they have no array leaves, so `eqx.partition` produces empty trainable trees and behavior is identical to the current static path.

For `FlexibleSigmoid`, `log_k` is a single scalar JAX array. The `softplus` + `expit` composition is fully differentiable, so gradient descent in the calibration block will optimize `log_k` alongside `α` and `logθ` with no additional machinery.

**`isinstance` checks inside JIT**: The `_ge_ridge_penalty` function uses `isinstance(model.global_epistasis, FlexibleSigmoid)`. This is safe because equinox preserves Python types during JAX tracing — the branch is resolved at trace time, not runtime. This is standard equinox practice. Do **not** replace with `jax.lax.cond`, which would require both branches to have the same pytree structure.

### Runtime overhead

Expected runtime overhead is negligible: one additional scalar parameter (`log_k`) in the calibration block gradient computation. The calibration block is already the cheapest block (only α and logθ), so adding one scalar does not materially affect total fit time.

### Verbose diagnostics

When `verbose=True`, the calibration block currently prints α and logθ per condition at each iteration. When `FlexibleSigmoid` is in use, also print the learned steepness `k` after each calibration block update:

```python
if verbose and isinstance(model.global_epistasis, FlexibleSigmoid):
    print(f"  k = {float(model.global_epistasis.k):.4f}")
```

### Convergence criterion

The calibration block already checks convergence via `state_calibration.error < tol`. Since `log_k` is optimized in the same block, its convergence is automatically included. No separate convergence check is needed.

### Initialization

`log_k` is initialized to `jnp.log(jnp.exp(1.0) - 1.0) ≈ 0.5413`, which gives `k = softplus(log_k) = 1.0` exactly. This ensures the model starts from the standard sigmoid, so the `ge_ridge` penalty has zero cost at initialization and the optimizer only moves `k` away from 1.0 when the data provides sufficient evidence.

## Testing Strategy

- **Unit tests for FlexibleSigmoid**:
  - `FlexibleSigmoid()(x)` with default `log_k=0` produces values in (0, 1)
  - `FlexibleSigmoid(log_k=large_value)` produces a steep step function
  - `k` property is always positive
  - Gradient of `log_k` is nonzero for non-trivial inputs
- **Backwards compatibility tests**:
  - `ge_type="Sigmoid"` produces identical results to current behavior
  - `ge_type="Identity"` produces identical results to current behavior
  - Models serialized with `pickle` (not `eqx.tree_serialise_leaves`) continue to load, since `Identity` and `Sigmoid` have no array leaves and their pytree structure is unchanged
- **Integration test for fitting**:
  - Fit with `FlexibleSigmoid` on synthetic data generated with true `k = 1.5`
  - Verify `model.steepness` recovers `k` within 20% relative error (i.e., `|k_fit - 1.5| / 1.5 < 0.2`)
  - Verify shift RMSE vs. ground truth decreases by at least 10% compared to fixed `Sigmoid` fit on the same data
- **Edge cases**:
  - `ge_ridge=0` (no regularization on k) — should still converge
  - Very large `ge_ridge` — k should stay near 1.0, recovering standard sigmoid behavior
  - `FlexibleSigmoid` with `Identity`-like data (k → 0 limit)

## Documentation Updates

- [ ] Update docstrings for `Model.__init__` (new `ge_type` option)
- [ ] Update docstrings for `Model.fit` (new `ge_ridge` parameter)
- [ ] Add `steepness` property to Model API docs
- [ ] Update `ge_type` Literal type annotation
- [ ] Add example in Sphinx docs showing FlexibleSigmoid usage

## Dependencies

No new dependencies. `jax.nn.softplus` is already available in the existing JAX dependency.

### Alternative 1: Learnable midpoint `m` and asymptotes `L`, `U`

Full generalized logistic: `g(x) = L + (U - L) / (1 + exp(-k * (x - m)))`

**Why not chosen:** `m` is redundant with `β0` (both shift the operating point), `L` and `U` are redundant with `α` (both scale the output range). Adding these parameters creates identifiability problems without adding expressiveness that isn't already captured by existing parameters. A single `k` is the minimal extension that adds genuinely new capability.

### Alternative 2: Per-condition steepness

Allow each condition to have its own `k_c`.

**Why not chosen:** Increases overfitting risk and partially duplicates `α`'s role. The global epistasis function is conceptually a shared property of the assay, not the biological condition. Per-condition differences should be captured by `α` and shifts. Can be revisited as future work if shared `k` proves insufficient.

### Alternative 3: Monotonic spline

Use a monotonic cubic spline with learnable knot values.

**Why not chosen:** Significantly more complex implementation (knot placement, monotonicity constraints via reparameterization or projection, choosing number of knots). Better suited as a follow-up if FlexibleSigmoid proves the concept. The spline would also require more careful regularization to avoid overfitting.

## Success Criteria

- [ ] `ge_type="FlexibleSigmoid"` fits successfully on spike DMS data (both replicates converge)
- [ ] Learned `k` differs from 1.0 by at least 0.1 (i.e., `|k - 1| > 0.1`), indicating the standard sigmoid is suboptimal for this data
- [ ] Delta replicate shift R² (Pearson) increases by at least 0.05 compared to fixed `Sigmoid` with the same regularization settings (baseline: R² = 0.28 at fusionreg=4.0)
- [ ] On synthetic data with known true `k ≠ 1`, `FlexibleSigmoid` recovers `k` within 20% relative error
- [ ] Existing `ge_type="Sigmoid"` and `ge_type="Identity"` produce bit-identical results to current behavior (same random seed, same parameters)
- [ ] All existing tests pass without modification

## Future Work

- Per-condition steepness (if shared `k` proves insufficient)
- Monotonic spline global epistasis (maximum flexibility)
- Asymmetric sigmoid (different steepness for positive vs. negative latent phenotype)
- Learned upper/lower asymptotes with appropriate identifiability constraints
